src/containers/tokens/components/bonding-curve-form/index.js

```diff
@@ -0,0 +1,422 @@
+
+  var count = 0
+  var dot = false
+  for (var i = 0; i < n.length; i++) {
+    if (n[i] === '.') dot = true
+    else if (n[i] !== '0' || count > 0)
+      // Past leading zeros.
+      count += 1
+
+    if (count >= 4 && dot) break
+  }
+  // Remove trailing dot.
+  if (n[i] === '.') i -= 1
+
+  return n.slice(0, i + 1)
+}
+
+class BondingCurveForm extends PureComponent {
+  static propTypes = {
+    buyPNKFromBondingCurveFormIsInvalid: PropTypes.bool.isRequired,
+    sellPNKToBondingCurveFormIsInvalid: PropTypes.bool.isRequired,
+    submitBuyPNKFromBondingCurveForm: PropTypes.func.isRequired,
+    submitSellPNKToBondingCurveForm: PropTypes.func.isRequired,
+    buyPNKFromBondingCurve: PropTypes.func.isRequired,
+    sellPNKToBondingCurve: PropTypes.func.isRequired,
+    approvePNKToBondingCurve: PropTypes.func.isRequired,
+    bondingCurveTotals:
+      bondingCurveSelectors.bondingCurveTotalsShape.isRequired,
+    inputETH: PropTypes.string,
+    inputPNK: PropTypes.string,
+    approveTransactionProgress: PropTypes.string.isRequired
+  }
+
+  static defaultProps = { inputETH: '0', inputPNK: '0' }
+
+  estimatePNK() {
+    const { inputETH, bondingCurveTotals } = this.props
+    const PNK = weiBNToDecimalString(
+      estimatePNK(
+        inputETH,
+        bondingCurveTotals.data.totalETH,
+        bondingCurveTotals.data.totalPNK
+      )
+    )
+    if (PNK === '0') return '0'
+    else return round(PNK)
+  }
+
+  buyPrice() {
+    const { inputETH, bondingCurveTotals } = this.props
+    const PNK = estimatePNK(
+      inputETH,
+      bondingCurveTotals.data.totalETH,
+      bondingCurveTotals.data.totalPNK
+    )
+    if (PNK === '0')
+      return round(
+        toFloat(
+          bondingCurveTotals.data.totalPNK,
+          bondingCurveTotals.data.totalETH
+        ).toString()
+      )
+    else
+      return round(
+        toFloat(toBN(PNK), decimalStringToWeiBN(inputETH)).toString()
+      )
+  }
+
+  estimateETH() {
+    const { inputPNK, bondingCurveTotals } = this.props
+    const ETH = weiBNToDecimalString(
+      estimateETH(
+        inputPNK,
+        bondingCurveTotals.data.totalETH,
+        bondingCurveTotals.data.totalPNK
+      )
+    )
+    if (ETH === '0') return '0'
+    else return round(ETH)
+  }
+
+  sellPrice() {
+    const { inputPNK, bondingCurveTotals } = this.props
+    const ETH = estimateETH(
+      inputPNK,
+      bondingCurveTotals.data.totalETH,
+      bondingCurveTotals.data.totalPNK
+    )
+    if (ETH === '0')
+      return round(
+        toFloat(
+          bondingCurveTotals.data.totalPNK,
+          bondingCurveTotals.data.totalETH
+        ).toString()
+      )
+    else
+      return round(
+        toFloat(decimalStringToWeiBN(inputPNK), toBN(ETH)).toString()
+      )
+  }
+
+  unlock = event => {
+    const { approvePNKToBondingCurve } = this.props
+    const APPROVE_PNK_AMOUNT = '100000000'
+
+    approvePNKToBondingCurve(
+      decimalStringToWeiBN(APPROVE_PNK_AMOUNT).toString()
```

Comment:
Why convert back to string?

Reply:
`decimalStringToWeiBN`, because it uses the `ethjs` library, returns a `bn.js` object. Down the call chain there is web3 which takes either `bignumber.js` object or string. So I convert the`bn.js` object to string here.